addons/account/models/product.py

```diff
@@ -118,6 +118,31 @@ def _check_uom_not_in_invoice(self):
                 "If you want to change its Unit of Measure, please archive this product and create a new one."
             ))
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        # If no company was set for the product, the product will be available for all companies and therefore should
+        # have the default taxes of the other companies as well. sudo() is used since we're going to need to fetch all
+        # the other companies default taxes which the user may not have access to.
+        other_companies = self.env['res.company'].sudo().search([('id', '!=', self.env.company.id)])
+        if other_companies:
+            for vals in vals_list:
+                if not vals.get('company_id'):
+                    default_taxes_ids = [company.account_sale_tax_id.id for company in other_companies if company.account_sale_tax_id]
+                    default_supplier_taxes_ids = [company.account_purchase_tax_id.id for company in other_companies if company.account_purchase_tax_id]
+
+                    if self._context.get('default_taxes_id', True):
+                        if 'taxes_id' in vals:
+                            vals['taxes_id'] = vals['taxes_id'] + [Command.link(tax_id) for tax_id in default_taxes_ids] if vals['taxes_id'] else [Command.set(default_taxes_ids)]
+                        else:
+                            vals['taxes_id'] = [Command.set(default_taxes_ids + self.env.company.account_sale_tax_id.ids)]
+
+                    if self._context.get('default_supplier_taxes_id', True):
+                        if 'supplier_taxes_id' in vals:
+                            vals['supplier_taxes_id'] = vals['supplier_taxes_id'] + [Command.link(tax_id) for tax_id in default_supplier_taxes_ids] if vals['supplier_taxes_id'] else [Command.set(default_supplier_taxes_ids)]
+                        else:
+                            vals['supplier_taxes_id'] = [Command.set(default_supplier_taxes_ids + self.env.company.account_purchase_tax_id.ids)]
+        return super(ProductTemplate, self.sudo()).create(vals_list)
```

Comment:
To me, you could simplify this a lot by creating the product first, then add the extra taxes if necessary.
```suggestion
@api.model_create_multi
def create(self, vals_list):
    products = super().create(vals_list)

    if products.company_id:
        return products

    # If no company was set for the product, the product will be available for all companies and therefore should
    # have the default taxes of the other companies as well. sudo() is used since we're going to need to fetch all
    # the other companies default taxes which the user may not have access to.
    other_companies = self.env['res.company'].sudo().search([])
    if not other_companies:
        return products

    default_customer_taxes = other_companies.filtered('account_sale_tax_id').account_sale_tax_id
    default_supplier_taxes = other_companies.filtered('account_purchase_tax_id').account_purchase_tax_id

    for product in products:
        if product.taxes_id:
            product_customer_taxes = default_customer_taxes - product.company_id.account_sale_tax_id
        else:
            product_customer_taxes = default_customer_taxes
        product.taxes_id = [Command.link(tax.id) for tax in product_customer_taxes]

        if product.supplier_taxes_ids:
            product_supplier_taxes = default_supplier_taxes - product.company_id.account_purchase_tax_id
        else:
            product_supplier_taxes = default_supplier_taxes
        product.supplier_taxes_ids = [Command.link(tax.id) for tax in product_supplier_taxes]

    return products
```